spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/gson/GsonAutoConfiguration.java

```diff
@@ -16,28 +16,176 @@
 
 package org.springframework.boot.autoconfigure.gson;
 
+import com.google.gson.FieldNamingPolicy;
 import com.google.gson.Gson;
 
+import com.google.gson.GsonBuilder;
+import com.google.gson.LongSerializationPolicy;
 import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+
+import org.springframework.context.annotation.Primary;
+import org.springframework.core.Ordered;
+import org.springframework.util.CollectionUtils;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
+import java.util.List;
 
 /**
  * {@link EnableAutoConfiguration Auto-configuration} for Gson.
  *
  * @author David Liu
+ * @author Ivan Golovko
  * @since 1.2.0
  */
 @Configuration
 @ConditionalOnClass(Gson.class)
 public class GsonAutoConfiguration {
 
-	@Bean
-	@ConditionalOnMissingBean
-	public Gson gson() {
-		return new Gson();
-	}
+    @Configuration
+    @ConditionalOnClass({Gson.class})
+    static class GsonConfiguration {
+
+        @Bean
+        @Primary
+        @ConditionalOnMissingBean(Gson.class)
+        public Gson gson(GsonBuilder gsonBuilder) {
+            return gsonBuilder.create();
+        }
+    }
+
+    @Configuration
+    @ConditionalOnClass({Gson.class})
+    static class GsonBuilderConfiguration {
+
+        @Bean
+        @ConditionalOnClass({Gson.class})
+        public GsonBuilder gsonBuilder(
+                List<GsonBuilderCustomizer> customizers
+        ) {
+            final GsonBuilder gsonBuilder = new GsonBuilder();
+            customizers.forEach(c -> c.customize(gsonBuilder));
+            return gsonBuilder;
+        }
+
+    }
+
+
+    @Configuration
+    @ConditionalOnClass({Gson.class})
+    @EnableConfigurationProperties(GsonProperties.class)
+    static class GsonBuilderCustomizerConfiguration {
+
+        @Bean
+        public StandardGsonBuilderCustomizer standardGsonBuilderCustomizer(
+                GsonProperties gsonProperties
+        ) {
+            return new StandardGsonBuilderCustomizer(gsonProperties);
+        }
+
+        private static final class StandardGsonBuilderCustomizer
+                implements GsonBuilderCustomizer, Ordered {
+
+            private final GsonProperties properties;
+
+            public StandardGsonBuilderCustomizer(GsonProperties properties) {
+                this.properties = properties;
+            }
```

Comment:
That is a lot of indirection, I am not sure that's actually required. Why don't you initialize the builder from `GsonProperties`? It could creates the instance, apply the default and then return that and all you'd have to do is apply the customizers. That would be 3 lines of code that can be in the existing `Gson` bean method.